text/2021-02-24-coprocessor-plugin.md

```diff
@@ -0,0 +1,145 @@
+# Coprocessor Plugin
+
+## Summary
+
+Add a general and plugable coprocessor framework for RawKV mode.
+
+## Motivation
+
+TiKV is the storage component in the TiDB ecosystem, however, the distribution computation principle suggests that computation should be as close to the data source as possible. Therefore, TiKV has embedded a subset of the TiDB executor framework to push down some computation tasks when applicable.
+
+But TiKV's capability should be far beyond that, as many distributed components can be built on top of TiKV, such as cache, full text search engine, gragh database and NoSQL database. And same as TiDB, these product will also like to push down specific computation to TiKV, which requires the coprocessor to be customizable, aka plugable.
+
+## Detailed design
+
+### Dynamic vs statically
+
+Generally, there are two strategies to build a plugin framework: dynamically and statically, which means to load the plugin on startup or to embed in the binary on compilation.
+
+![Plugin Arch](../media/plugin-arch.png)
+
+They have both pros and cons:
+
+| Static | Dynamic |
+| -- | -- |
+| ◯ High performance | X Relatively slower |
+| ◯ Easy to deploy | X Complexify the deploy process |
+| X Build the entire TiKV | ◯ Easy to build |
+| X Build very slow | ◯ Build fast |
+| X Hard to debug | ◯ Easy to debug |
+
+Ideally, we'd like to develop a plugin in dynamic mode, and eventually, to distribute the statically linked one in release. So there is an interesting research area that designing a plugin framework that can write the plugin once, and compile to the dynamic one and the static one with duplicate code. Anyway, this is an ambitious target but sort of out of scope, so we may explore the possibility when marching on.
+
+So initially, in this RFC, we'll only focus on the dynamic plugin framework that works in RawKV mode.
+
+### Web Assembly
+
+Web Assembly is chosen to host the dynamic plugin. There was alternatives like dynamic lib, lua and bpf. So in this section, I'll explain the tradeoff between them and why WASM is the most appropriate choice for TiKV:
+
+- Dynamic Library
+
+    Previously, a proof-of-concept plugin experiment [[repo]](https://github.com/andylokandy/plugin) is done. As a result, we found that Rust's unstable ABI is a risk in safety. It's hard to guarantee that the TiKV and the plugin has absolutely the same ABI version, and is also hard to debug such a plugin.
+
+- Lua
+
+    TOO SLOW
+
+- eBPF
+
+    Berkeley Packet Filter has been experimented in Hackthon. It requires the plugin to be written in C, which is not capable to migrate `tidb_query` eventually. And also, eBPF is not turning-complete, which is unacceptable.
```

Comment:
Berkeley -> Berkerley, Hackthon -> Hackathon, turning-complete -> turing-complete

Reply:
It's `Berkeley`: https://en.wikipedia.org/wiki/Berkeley_Packet_Filter